### PR TITLE
[SYCL] Check that apps using sycl also use dynamic C++ RT even withou…

### PR DESCRIPTION
**Pull request by v-klochkov**
…t -fsycl

With -fsycl switch the dynamic runtime is used.
The patches (#2478, #2480, #2497 implemented that with -fsycl.

This patch adds check that dynamic runtime is used when -fsycl is
not used. That is done by a static_assert in sycl/stl.hpp.

Also, lots of LIT tests had to be changed to comply with the new
requirement (apps must use dynamic C++ RT with use SYCL).

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

### CODE REVIEW

**Review by alexbatashev — DISMISSED** [review]

File: sycl/test/basic_tests/buffer/buffer.cpp [hunk 1]
```diff
@@ -1,5 +1,3 @@
-// RUN: %clangxx %s -o %t1.out -lsycl -I %sycl_include
```
Comment (vladimirlaz) [inline sycl/test/basic_tests/buffer/buffer.cpp:1]: Did we prohibit use of the compile to build C++ code? Otherwise we do not need to remove this code.

File: sycl/test/basic_tests/implicit_conversion_error.cpp [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// RUN: %clangxx -fsyntax-only -Xclang -verify  %s -I %sycl_include -Xclang -verify-ignore-unexpected=note,warning
```
Comment (vladimirlaz) [inline sycl/test/basic_tests/implicit_conversion_error.cpp:1]: Do we really need to change this test as it is compile-time check?

File: sycl/test/basic_tests/buffer/buffer.cpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/test/basic_tests/buffer/buffer.cpp:1]: This command compiles a test with static C++ RT on Windows, which causes an assert in stl.hpp
(the assert is added in this PR).

Comment (v-klochkov) [reply sycl/test/basic_tests/buffer/buffer.cpp:1]: Vladimir, Do you have good ideas on how to allow such compilation on Linux (where default C++ RT is dynamic) and not allow on Windows without -fsycl (where default C++ RT is static)?

File: sycl/test/basic_tests/implicit_conversion_error.cpp (hunk 2 above)
Comment (v-klochkov) [reply sycl/test/basic_tests/implicit_conversion_error.cpp:1]: The new static_assert at stl.hpp is emitted on Windows when a) user does not use -fsycl and b) when such compilation would potentially lead to linking with static RT.
I will check if there is any difference in compilation macros with/with -fsyntax-only and if that difference usable in stl.hpp to not give an error with -fsyntax-only.

File: sycl/test/basic_tests/buffer/buffer.cpp (hunk 1 above)
Comment (vladimirlaz) [reply sycl/test/basic_tests/buffer/buffer.cpp:1]: We have OS specific placehoders for OpenCL CPU: https://github.com/intel/llvm/blob/sycl/sycl/test/lit.cfg.py#L144
You can have some general LINUX placeholder to run the compilation only on Linux.

But it is not perfect solution so I will not insist on implementing it.

**Review by alexbatashev — DISMISSED** [review]

File: sycl/include/CL/sycl/stl.hpp [hunk 3]
```diff
@@ -22,34 +22,52 @@
 __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/stl.hpp:24]: Temporary put this PR on hold to wait for more elaborate testing results.

**Review by romanovvlad — CHANGES_REQUESTED** [review]

File: sycl/include/CL/sycl/stl.hpp (hunk 3 above)
Comment (bader) [reply sycl/include/CL/sycl/stl.hpp:24]: @romanovvlad, any updates on testing results?

Comment (romanovvlad) [reply sycl/include/CL/sycl/stl.hpp:24]: I put this comment to mark that PR cannot be committed yet in agreement with Slava.
@v-klochkov  Could you please help answering?

**Comment by bader** [comment]
@v-klochkov, could you resolve merge conflicts and address comments from @romanovvlad, please?

File: sycl/include/CL/sycl/stl.hpp [hunk 4]
```diff
@@ -22,34 +22,52 @@
...
+template <class T, class Deleter = std::default_delete<T>>
+using unique_ptr_class = std::unique_ptr<T, Deleter>;
 
- template <class T>
- using weak_ptr_class = std::weak_ptr<T>;
+template <class T> using shared_ptr_class = std::shared_ptr<T>;
 
- template <class T>
- using hash_class = std::hash<T>;
+template <class T> using weak_ptr_class = std::weak_ptr<T>;
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/stl.hpp:60]: Please remove unrelated formatting changes here and below.

File: sycl/include/CL/sycl/stl.hpp [hunk 5]
```diff
@@ -22,34 +22,52 @@
...
 
- template < class T, class Alloc = std::allocator<T> >
- using vector_class = std::vector<T, Alloc>;
+#if defined(_WIN32) && !defined(__SYCL_DEVICE_ONLY__)
+namespace detail {
+// SYCL library is designed such a way that STL objects cross DLL boundary,
+// which is not guaranteed to work and considered not safe in general.
+// Only using same dynamic C++ runtime library for sycl[d].dll and for
+// the application using sycl[d].dll is guaranteed to work properly.
+inline constexpr bool isMSVCDynamicCXXRuntime() {
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/stl.hpp:31]: `constexpr` implies `inline`

```suggestion
constexpr bool isMSVCDynamicCXXRuntime() {
```

File: sycl/test/basic_tests/context.cpp [hunk 6]
```diff
@@ -1,13 +1,7 @@
...
 // RUN: %RUN_ON_HOST %t.out
 
-//==--------------- context.cpp - SYCL context test ------------------------==//
-//
-// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
-// See https://llvm.org/LICENSE.txt for license information.
-// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-//
-//===----------------------------------------------------------------------===//
+// This test performs basic check of the SYCL context class.
```
Comment (alexbatashev) [inline sycl/test/basic_tests/context.cpp:4]: These changes seem to be irrelevant

File: sycl/include/CL/sycl/stl.hpp (hunk 4 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/stl.hpp:60]: That was already discussed previously - I did that formatting because clang-format failed without it. Somehow it required changes for lines of code that were just adjacent to those that I changed.

File: sycl/test/basic_tests/context.cpp (hunk 6 above)
Comment (v-klochkov) [reply sycl/test/basic_tests/context.cpp:4]: Yes, but these changes are not wrong.

File: sycl/test/basic_tests/buffer/buffer.cpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/test/basic_tests/buffer/buffer.cpp]: I did not find any good way to detect if the compilation runs with -fsyntax-only or without it.

**Review by kbobrovs — APPROVED** [review]
sycl/test/esimd/vadd.cpp LGTM.

**Comment by v-klochkov** [comment]
3 or 4 failed tests from the Jenkins/Precommit use separate compilation, e.g.:
  https://github.com/intel/llvm-test-suite/blob/intel/SYCL/Config/kernel_from_file.cpp
  https://github.com/intel/llvm-test-suite/blob/intel/SYCL/DeviceLib/separate_compile_test.cpp

Such usage model caused the hit of the new static_assert (added in this PR). 
It might be useful to add "-fsycl-host-only" option (similar to existing "-fsycl-device-only") in long terms if DPC++ keeps supporting separate device/host compilation.

In short terms it makes sense replacing "static_assert()" with '#pragma message()".

**Review by alexbatashev — APPROVED** [review]
LGTM

File: sycl/test/esimd/vadd.cpp [hunk 7]
```diff
@@ -1,5 +1,4 @@
-// TODO ESIMD enable host device under -fsycl
-// RUN: %clangxx -I %sycl_include %s -o %t.out -lsycl
+// RUN: %clangxx -fsycl %s -o %t.out
```
Comment (DenisBakhvalov) [inline sycl/test/esimd/vadd.cpp:1]: Sorry for not catching it, but there is a regression for this test. BTW, I'm not sure how it slipped through the CI.
This test fails with:
```
InvalidFunctionCall: Unexpected llvm intrinsic:
 llvm.genx.local.id.v3i32 [Src: /iusers/dbakhval/github/llvm/llvm-spirv/lib/SPIRV/SPIRVWriter.cpp:2782 false ]
```
As of today, we still require `-fsycl-explicit-simd` flag for ESIMD code. I submitted #3457 to fix this.

Comment (bader) [reply sycl/test/esimd/vadd.cpp:1]: > BTW, I'm not sure how it slipped through the CI.

This casts shadow on CI validation and ESIMD feature validation in particular. @DenisBakhvalov, could you investigate why this regression happened, please? Could it be a conflict between commits?
